File: crates/skia-rs-python/examples/basic_drawing.py
#!/usr/bin/env python3
"""
Basic drawing example for skia-rs Python bindings.

This example demonstrates:
- Creating a surface
- Drawing basic shapes (rectangles, circles, lines)
- Using different paint styles (fill, stroke)
- Working with paths
- Exporting to PNG (requires PIL/Pillow)
"""

import skia_rs

# Optional: for saving to PNG
try:
    import numpy as np
    from PIL import Image
    HAS_PIL = True
except ImportError:
    HAS_PIL = False
    print("Note: PIL/Pillow not installed. Cannot save PNG.")


def main():
    # Create a 800x600 surface
    width, height = 800, 600
    surface = skia_rs.Surface(width, height)

    # Clear with white background
    surface.clear(skia_rs.Colors.WHITE)

    # Create a filled rectangle paint
    fill_paint = skia_rs.Paint()
    fill_paint.color = skia_rs.rgb(100, 150, 200)  # Light blue
    fill_paint.anti_alias = True

    # Draw a filled rectangle
    surface.draw_rect(50, 50, 250, 200, fill_paint)

    # Create a stroke paint
    stroke_paint = skia_rs.Paint()
    stroke_paint.style = "stroke"
    stroke_paint.stroke_width = 4.0
    stroke_paint.color = skia_rs.Colors.RED
    stroke_paint.anti_alias = True

    # Draw a stroked circle
    surface.draw_circle(400, 150, 80, stroke_paint)

    # Draw an oval
    oval_paint = skia_rs.Paint()
    oval_paint.color = skia_rs.argb(180, 0, 200, 100)  # Semi-transparent green
    surface.draw_oval(500, 50, 750, 200, oval_paint)

    # Draw lines
    line_paint = skia_rs.Paint()
    line_paint.style = "stroke"
    line_paint.stroke_width = 2.0
    line_paint.color = skia_rs.Colors.BLUE

    for i in range(10):
        y = 250 + i * 10
        surface.draw_line(50, y, 350, y + 50, line_paint)

    # Build and draw a path (a star shape)
    builder = skia_rs.PathBuilder()
    cx, cy = 600, 400
    outer_r, inner_r = 100, 40

    import math
    for i in range(5):
        angle_outer = math.radians(i * 72 - 90)
        angle_inner = math.radians(i * 72 + 36 - 90)

        ox = cx + outer_r * math.cos(angle_outer)
        oy = cy + outer_r * math.sin(angle_outer)
        ix = cx + inner_r * math.cos(angle_inner)
        iy = cy + inner_r * math.sin(angle_inner)

        if i == 0:
            builder.move_to(ox, oy)
        else:
            builder.line_to(ox, oy)
        builder.line_to(ix, iy)
    builder.close()

    star_path = builder.build()

    star_paint = skia_rs.Paint()
    star_paint.color = skia_rs.Colors.YELLOW
    surface.draw_path(star_path, star_paint)

    # Stroke the star outline
    star_stroke = skia_rs.Paint()
    star_stroke.style = "stroke"
    star_stroke.stroke_width = 2.0
    star_stroke.color = skia_rs.rgb(200, 150, 0)  # Dark yellow
    surface.draw_path(star_path, star_stroke)

    # Draw a rounded rectangle
    rounded_paint = skia_rs.Paint()
    rounded_paint.color = skia_rs.argb(200, 150, 100, 200)  # Purple

    builder2 = skia_rs.PathBuilder()
    builder2.add_round_rect(50, 350, 300, 550, 20, 20)
    surface.draw_path(builder2.build(), rounded_paint)

    print(f"Created {width}x{height} image")
    print(f"Star path bounds: {star_path.bounds()}")
    print(f"Star contains center: {star_path.contains(600, 400)}")

    # Save to PNG if PIL is available
    if HAS_PIL:
        pixels = surface.pixels()
        arr = np.frombuffer(pixels, dtype=np.uint8).reshape(height, width, 4)
        img = Image.fromarray(arr, mode='RGBA')
        img.save('python_drawing.png')
        print("Saved to python_drawing.png")


if __name__ == "__main__":
    main()

File: crates/skia-rs-node/examples/basic_drawing.js
#!/usr/bin/env node
/**
 * Basic drawing example for skia-rs Node.js bindings.
 *
 * This example demonstrates:
 * - Creating a surface
 * - Drawing basic shapes (rectangles, circles, lines)
 * - Using different paint styles (fill, stroke)
 * - Working with paths
 * - Exporting to PNG (requires pngjs)
 */

const skia = require('../index.js');
const fs = require('fs');

// Optional: for saving to PNG
let PNG;
try {
    PNG = require('pngjs').PNG;
} catch (e) {
    console.log('Note: pngjs not installed. Cannot save PNG.');
    console.log('Run: npm install pngjs');
}

function main() {
    // Create a 800x600 surface
    const width = 800;
    const height = 600;
    const surface = new skia.Surface(width, height);

    // Clear with white background
    surface.clear(skia.WHITE);

    // Create a filled rectangle paint
    const fillPaint = new skia.Paint();
    fillPaint.setArgb(255, 100, 150, 200); // Light blue
    fillPaint.setAntiAlias(true);

    // Draw a filled rectangle
    surface.drawRect(50, 50, 250, 200, fillPaint);

    // Create a stroke paint
    const strokePaint = new skia.Paint();
    strokePaint.setStyle(1); // stroke
    strokePaint.setStrokeWidth(4.0);
    strokePaint.setColor(skia.RED);
    strokePaint.setAntiAlias(true);

    // Draw a stroked circle
    surface.drawCircle(400, 150, 80, strokePaint);

    // Draw an oval
    const ovalPaint = new skia.Paint();
    ovalPaint.setArgb(180, 0, 200, 100); // Semi-transparent green
    surface.drawOval(500, 50, 750, 200, ovalPaint);

    // Draw lines
    const linePaint = new skia.Paint();
    linePaint.setStyle(1); // stroke
    linePaint.setStrokeWidth(2.0);
    linePaint.setColor(skia.BLUE);

    for (let i = 0; i < 10; i++) {
        const y = 250 + i * 10;
        surface.drawLine(50, y, 350, y + 50, linePaint);
    }

    // Build and draw a path (a star shape)
    const builder = new skia.PathBuilder();
    const cx = 600, cy = 400;
    const outerR = 100, innerR = 40;

    for (let i = 0; i < 5; i++) {
        const angleOuter = (i * 72 - 90) * Math.PI / 180;
        const angleInner = (i * 72 + 36 - 90) * Math.PI / 180;

        const ox = cx + outerR * Math.cos(angleOuter);
        const oy = cy + outerR * Math.sin(angleOuter);
        const ix = cx + innerR * Math.cos(angleInner);
        const iy = cy + innerR * Math.sin(angleInner);

        if (i === 0) {
            builder.moveTo(ox, oy);
        } else {
            builder.lineTo(ox, oy);
        }
        builder.lineTo(ix, iy);
    }
    builder.close();

    const starPath = builder.build();

    const starPaint = new skia.Paint();
    starPaint.setColor(skia.YELLOW);
    surface.drawPath(starPath, starPaint);

    // Stroke the star outline
    const starStroke = new skia.Paint();
    starStroke.setStyle(1); // stroke
    starStroke.setStrokeWidth(2.0);
    starStroke.setArgb(255, 200, 150, 0); // Dark yellow
    surface.drawPath(starPath, starStroke);

    // Draw a rounded rectangle
    const roundedPaint = new skia.Paint();
    roundedPaint.setArgb(200, 150, 100, 200); // Purple

    const builder2 = new skia.PathBuilder();
    builder2.addRoundRect(50, 350, 300, 550, 20, 20);
    surface.drawPath(builder2.build(), roundedPaint);

    console.log(`Created ${width}x${height} image`);
    const bounds = starPath.getBounds();
    console.log(`Star path bounds: Rect(${bounds.left}, ${bounds.top}, ${bounds.right}, ${bounds.bottom})`);
    console.log(`Star contains center: ${starPath.contains(600, 400)}`);

    // Save to PNG if pngjs is available
    if (PNG) {
        const pixels = surface.getPixels();
        const png = new PNG({ width, height });
        png.data = pixels;
        const buffer = PNG.sync.write(png);
        fs.writeFileSync('node_drawing.png', buffer);
        console.log('Saved to node_drawing.png');
    }
}

main();
